### ISSUE TITLE
RFC: Customizable `wait()` behavior

### ISSUE THREAD

**Issue by nikomatsakis** [alexcrichton/futures-rs#360]
The existing `wait()` method *always* blocks the current thread. However, as the docs themselves state, this is very often not the thing you want to do:

> Note: This method is not appropriate to call on event loops or similar I/O situations because it will prevent the event loop from making progress (this blocks the thread). This method should only be called when it's guaranteed that the blocking work associated with this future will be completed by another thread.

In Rayon, I made the `wait()` method (when invoked on a `RayonFuture`, the only kind I have control of) simply `panic()` if it is used on a Rayon thread. I then added a second method, `rayon_wait()`, that will interoperate properly with Rayon's scheduler.

It seems to me it would make sense for the futures library to offer some kind of thread-local interface to customize what `wait()` does. The default would no doubt be the current behavior, but executors like Rayon could -- for their threads, or within their scope -- override this by setting an appropriate thread-local callback.

The main counterargument I see is that this might encourage people to write code that uses `wait()`, which I would consider largely an anti-pattern. The better thing to do is to create (and schedule) a derived future. In this case, I have to wonder if including `wait()` in the Futures interface *at all* makes sense, though.

**Comment by alexcrichton** [comment]
Makes sense to me! You could even imagine that this feature could be leveraged to implement coroutines-as-futures where `.wait()` just does the context switch away if it's not ready yet.

**Comment by alexcrichton** [comment]
This relates to some of the discussion @carllerche @aturon and I had about https://github.com/tokio-rs/tokio-core/issues/79 as well. I think this would also be quite useful for tokio-core clients. What I'm thinking is:

* At the beginning of your thread you create `Core` and customize the `wait` behavior to use this core.
* Otherwise in your program you just call `wait`. 

The calls to `wait` would run the event loop on the current thread, waiting for the passed in future to resolve.

**Comment by nikomatsakis** [comment]
@alexcrichton 

> Otherwise in your program you just call wait.

By "otherwise", do you mean "elsewhere"? i.e., once you have created a `Core` to customize the wait behavior, you can just call `wait`? If so, it seems reasonable, though I worry about footguns if users forget to create a `Core`. But I'm lacking context.

**Comment by alexcrichton** [comment]
Ah yes yeah my command of the English language isn't always there... 

In any case yeah, my idea is that you get used to just calling `wait` when you want to block, and it's automatically taken care of to run the event loop at that time. It's true though that this is a footgun-waiting-to-happen if you don't actually configure the behavior...

**Comment by carllerche** [comment]
IMO this would be good, especially since it would enable executors to specify that a call to wait on an event loop that doesn't support blocking is a panic.

Adding support for this as part of #436 would be fairly easy. Since `Unpark` is now an executor level value, there could be a hook added for waiting.

**Comment by alexcrichton** [comment]
Ok here's my idea of how to implement this: https://github.com/alexcrichton/futures-rs/commit/390f34bcd9617b658facae7b501d5d63b6e0aef2 and [here's a comment](https://github.com/alexcrichton/futures-rs/pull/436#issuecomment-301673150) for how it might look in tokio-core.

@nikomatsakis mind taking a look to see if that would be totally unreasonable for rayon?

**Comment by alexcrichton** [comment]
I have submitted a PR for this: https://github.com/alexcrichton/futures-rs/pull/487

**Comment by carllerche** [comment]
I believe this this is now a "won't fix" with the direction being taken in #645. Please re-open if I am wrong.
